pre-alignment-qc-wf/main.nf

```diff
@@ -0,0 +1,242 @@
+  take:  // update as needed
+    study_id
+    analysis_id
+    analysis_metadata
+    experiment_info_tsv
+    read_group_info_tsv
+    file_info_tsv
+    extra_info_tsv
+    sequencing_files
+
+
+  main:  // update as needed
+    // detect local mode or not
+    local_mode = false
+    if ((!analysis_metadata.startsWith("NO_FILE") || !experiment_info_tsv.startsWith("NO_FILE")) && sequencing_files.size() > 0){
+        local_mode = true
+        if (!params.publish_dir) {
+            exit 1, "You specified local sequencing data as input, please also set `params.publish_dir` to keep the output."
+        }
+        log.info "Run the workflow using local input sequencing data, results will be in: ${params.publish_dir}"
+
+        if (!analysis_metadata.startsWith("NO_FILE")) {
+            if (!experiment_info_tsv.startsWith("NO_FILE") ||
+                    !read_group_info_tsv.startsWith("NO_FILE") ||
+                    !file_info_tsv.startsWith("NO_FILE") ||
+                    !extra_info_tsv.startsWith("NO_FILE")
+            )  {
+                log.info "Use analysis metadata JSON as input, will ignore input: 'experiment_info_tsv', 'read_group_info_tsv', 'file_info_tsv', 'extra_info_tsv'"
+            }
+            analysis_metadata = file(analysis_metadata)
+        } else if (!experiment_info_tsv.startsWith("NO_FILE") &&
+                    !read_group_info_tsv.startsWith("NO_FILE") &&
+                    !file_info_tsv.startsWith("NO_FILE") &&
+                    !extra_info_tsv.startsWith("NO_FILE")
+            ) {
+            pGenExp(
+                file(experiment_info_tsv),
+                file(read_group_info_tsv),
+                file(file_info_tsv),
+                file(extra_info_tsv)
+            )
+            analysis_metadata = pGenExp.out.payload
+        } else {
+            exit 1, "To run the workflow using local inputs, please specify metadata in JSON using params.analysis_metadata or metadata in TSVs using params.experiment_info_tsv, params.read_group_info_tsv, params.file_info_tsv and params.extra_info_tsv"
+        }
+
+        sequencing_files = Channel.fromPath(sequencing_files)
+    } else if (study_id && analysis_id) {
+        // download files and metadata from song/score (analysis type: sequencing_experiment)
+        log.info "Run the workflow using input sequencing data from SONG/SCORE, alignment results will be uploaded to SONG/SCORE as well"
+        dnld(study_id, analysis_id)
+        analysis_metadata = dnld.out.analysis_json
+        sequencing_files = dnld.out.files
+    } else {
+        exit 1, "To use sequencing data from SONG/SCORE as input, please provide `params.study_id`, `params.analysis_id` and other SONG/SCORE params.\n" +
+            "Or please provide `params.analysis_metadata` (or `params.experiment_info_tsv`, `params.read_group_info_tsv`, `params.file_info_tsv` and `params.extra_info_tsv`) and `params.sequencing_files` from local files as input."
```

Comment:
Sequencing files also need to be size > 0

Reply:
Yes. that's correct.